src/main/java/net/openhft/chronicle/core/io/TracingReferenceCounted.java

```diff
@@ -230,11 +220,18 @@ public void throwExceptionIfReleased() throws ClosedIllegalStateException {
 
     @Override
     public void warnAndReleaseIfNotReleased() {
-        if (refCount() > 0) {
-            if (!unmonitored && !AbstractCloseable.DISABLE_DISCARD_WARNING) {
-                ExceptionHandler warn = AbstractCloseable.STRICT_DISCARD_WARNING ? Jvm.warn() : Slf4jExceptionHandler.WARN;
-                warn.on(type, "Discarded without being released by " + referencesAsString(), createdHere);
+        boolean runOnRelease = false;
+        synchronized (references) {
+            if (refCount() > 0) {
+                if (!unmonitored && !AbstractCloseable.DISABLE_DISCARD_WARNING) {
+                    Jvm.warn().on(type, "Discarded without being released by " + referencesAsString(), createdHere);
+                }
+                references.clear();
+                runOnRelease = true;
             }
+        }
+        // Run outside the synchronized block to avoid risk of deadlock
+        if (runOnRelease) {
```

Comment:
I changed warnAndRelease to actually clear the references, so any subsequent calls to reserve or release would fail. I did this with Tracing and Vanilla

Reply:
+1 The method name does imply that they should be released.